## Set Up

## Import libraries

In [56]:
import pandas as pd


### Import Config

In [57]:
import configparser

config = configparser.ConfigParser()
config.read("../../../env.ini")
data_home = config['DEFAULT']['data_home']
output_dir = config['DEFAULT']['output_dir']

In [58]:
data_home, output_dir

('/Users/jamessiegener/MSDS/DS5001/data',
 '/Users/jamessiegener/MSDS/DS5001/output')

In [59]:
text_file = f"{data_home}/gutenberg/pg161.txt"
csv_file = f"{output_dir}/austen-combo-TOKENS.csv"

In [60]:
OHCO = ['chap_num', 'para_num', 'sent_num', 'token_num']

## Import file into a dataframe

In [61]:
LINES = pd.DataFrame(open(text_file, 'r', encoding='utf-8-sig').readlines(), columns=['line_str'])
LINES.index.name = 'line_num'
LINES.line_str = LINES.line_str.str.replace(r'\n+', ' ', regex=True).str.strip()
LINES.sample(20)

line_str
line_num                                                   
40                                                         
6082      "Exert yourself, dear Marianne," she cried, "i...
3867      too.-- I assure you it was a great compliment ...
4064                                                       
72        property, could be but small.  Their mother ha...
7530      people of common prudence will do THAT; and wh...
754                                                        
11312     brought, by their united request, to consider ...
11107                                              her own.
5523      Colonel Brandon, who had a general invitation ...
6103                                    CAN have no grief."
6562      "Marianne is not well," said she.  "She has be...
2246      "But, my dear Marianne, as it has already expo...
5103                            estranged from each other."
7440                                                       
10182     looked very grave on Mr. Harris's report, and ...
10848     town, however, I blundered on Sir John, I beli...
4577      while the same period of time, spent on her si...
1283      from their hills; and the two girls set off to...
5537      Elinor, rejoiced to be assured of his being in...

## Extract Title

In [62]:
title = LINES.loc[0].line_str.replace('The Project Gutenberg EBook of ', '')
title

'Sense and Sensibility, by Jane Austen'

## Clip the Cruft

In [63]:
clip_pats = [
    r"\*\*\*\s*START OF (?:THE|THIS) PROJECT",
    r"\*\*\*\s*END OF (?:THE|THIS) PROJECT"
]

In [64]:
pat_a = LINES.line_str.str.match(clip_pats[0])
pat_b = LINES.line_str.str.match(clip_pats[1])

In [65]:
line_a = LINES.loc[pat_a].index[0] + 1
line_b = LINES.loc[pat_b].index[0] - 1

In [66]:
line_a, line_b

(20, 12666)

In [67]:
LINES = LINES.loc[line_a: line_b]

In [68]:
LINES.head(10)

line_str
line_num         
20               
21               
22               
23               
24               
25               
26               
27               
28               
29

In [69]:
LINES.tail(10)

line_str
line_num                                                   
12657                                                      
12658                                                      
12659                                                      
12660                                                      
12661                                                      
12662                                                      
12663                                                      
12664                                                      
12665     End of the Project Gutenberg EBook of Sense an...
12666

## Chunk by Chapter

### Find all chapter headers

In [70]:
chap_pat = r"^\s*(?:chapter|letter)\s+\d+"

In [71]:
chap_lines = LINES.line_str.str.match(chap_pat, case=False)

In [72]:
LINES.loc[chap_lines] 

line_str
line_num            
42         CHAPTER 1
196        CHAPTER 2
399        CHAPTER 3
561        CHAPTER 4
756        CHAPTER 5
858        CHAPTER 6
986        CHAPTER 7
1112       CHAPTER 8
1244       CHAPTER 9
1448      CHAPTER 10
1665      CHAPTER 11
1816      CHAPTER 12
1997      CHAPTER 13
2281      CHAPTER 14
2440      CHAPTER 15
2718      CHAPTER 16
2945      CHAPTER 17
3153      CHAPTER 18
3331      CHAPTER 19
3632      CHAPTER 20
3913      CHAPTER 21
4214      CHAPTER 22
4532      CHAPTER 23
4767      CHAPTER 24
5001      CHAPTER 25
5197      CHAPTER 26
5454      CHAPTER 27
5732      CHAPTER 28
5883      CHAPTER 29
6324      CHAPTER 30
6628      CHAPTER 31
7004      CHAPTER 32
7278      CHAPTER 33
7601      CHAPTER 34
7888      CHAPTER 35
8152      CHAPTER 36
8456      CHAPTER 37
8900      CHAPTER 38
9205      CHAPTER 39
9408      CHAPTER 40
9706      CHAPTER 41
9977      CHAPTER 42
10155     CHAPTER 43
10490     CHAPTER 44
11060     CHAPTER 45
11278     CHAPTER 46
11571     CHAPTER 47
11838     CHAPTER 48
11986     CHAPTER 49
12410     CHAPTER 50

### Assign numbers to chapters

In [73]:
LINES.loc[chap_lines, 'chap_num'] = [i+1 for i in range(LINES.loc[chap_lines].shape[0])]

In [74]:
LINES.loc[chap_lines]

line_str  chap_num
line_num                      
42         CHAPTER 1       1.0
196        CHAPTER 2       2.0
399        CHAPTER 3       3.0
561        CHAPTER 4       4.0
756        CHAPTER 5       5.0
858        CHAPTER 6       6.0
986        CHAPTER 7       7.0
1112       CHAPTER 8       8.0
1244       CHAPTER 9       9.0
1448      CHAPTER 10      10.0
1665      CHAPTER 11      11.0
1816      CHAPTER 12      12.0
1997      CHAPTER 13      13.0
2281      CHAPTER 14      14.0
2440      CHAPTER 15      15.0
2718      CHAPTER 16      16.0
2945      CHAPTER 17      17.0
3153      CHAPTER 18      18.0
3331      CHAPTER 19      19.0
3632      CHAPTER 20      20.0
3913      CHAPTER 21      21.0
4214      CHAPTER 22      22.0
4532      CHAPTER 23      23.0
4767      CHAPTER 24      24.0
5001      CHAPTER 25      25.0
5197      CHAPTER 26      26.0
5454      CHAPTER 27      27.0
5732      CHAPTER 28      28.0
5883      CHAPTER 29      29.0
6324      CHAPTER 30      30.0
6628      CHAPTER 31      31.0
7004      CHAPTER 32      32.0
7278      CHAPTER 33      33.0
7601      CHAPTER 34      34.0
7888      CHAPTER 35      35.0
8152      CHAPTER 36      36.0
8456      CHAPTER 37      37.0
8900      CHAPTER 38      38.0
9205      CHAPTER 39      39.0
9408      CHAPTER 40      40.0
9706      CHAPTER 41      41.0
9977      CHAPTER 42      42.0
10155     CHAPTER 43      43.0
10490     CHAPTER 44      44.0
11060     CHAPTER 45      45.0
11278     CHAPTER 46      46.0
11571     CHAPTER 47      47.0
11838     CHAPTER 48      48.0
11986     CHAPTER 49      49.0
12410     CHAPTER 50      50.0

In [75]:
LINES.sample(10)

line_str  chap_num
line_num                                                             
3003      "About eighteen hundred or two thousand a year...       NaN
4780      "Offended me! How could you suppose so? Believ...       NaN
9696      himself, he thinks that nobody else can marry ...       NaN
4930      more for the happiness of both of you,' I shou...       NaN
7025      she could not bring herself to speak of what s...       NaN
10597     me before she ought to have done it.  But she ...       NaN
10599                                                             NaN
597       the world of his goodness and sense.  I think ...       NaN
3995      whatever her ladyship was doing, if she happen...       NaN
9089                                                              NaN

In [76]:
LINES.chap_num = LINES.chap_num.ffill()

In [77]:
LINES.sample(10)

line_str  chap_num
line_num                                                             
9568                                                             40.0
5900      longer than while she spoke, and was immediate...      29.0
12533     marrying privately without his mother's consen...      50.0
8828      you have a regard for, Mrs. Jennings.  We all ...      37.0
7994      it was so.  They all looked exceedingly foolis...      35.0
8740               affair, and bring them news of his wife.      37.0
10312     servant with a message to Mr. Harris, and an o...      43.0
11450                                       expose me to"--      46.0
92        all the value of all the attention which, for ...       1.0
3619                                                             19.0

In [78]:
LINES.head(20)

line_str  chap_num
line_num                                 
20                                    NaN
21                                    NaN
22                                    NaN
23                                    NaN
24                                    NaN
25                                    NaN
26                                    NaN
27                                    NaN
28                                    NaN
29                                    NaN
30                                    NaN
31                                    NaN
32                                    NaN
33        SENSE AND SENSIBILITY       NaN
34                                    NaN
35               by Jane Austen       NaN
36                                    NaN
37                       (1811)       NaN
38                                    NaN
39                                    NaN

In [79]:
LINES = LINES.dropna(subset=['chap_num']) 
LINES = LINES.loc[~chap_lines] 
LINES.chap_num = LINES.chap_num.astype('int')

In [80]:
LINES.sample(10)

line_str  chap_num
line_num                                                             
9894      left to improve her acquaintance with Robert, ...        41
9722      own all their obligation to her, and openly de...        41
11861     place in which so much conspired to give her a...        48
12177                                                              49
8906      consolation, beyond the consciousness of doing...        38
10240     attended her every day, still talked boldly of...        43
5933      Willoughby, which appeared to her a very good ...        29
3283      alone--and tomorrow you must absolutely dine w...        18
4937                       much for an indifferent person."        24
8050      should be checked by Lucy's unwelcome presence...        35

In [81]:
CHAPS = LINES.groupby(OHCO[:1])\
    .line_str.apply(lambda x: '\n'.join(x))\
    .to_frame('chap_str')

In [82]:
CHAPS.head(10)

chap_str
chap_num                                                   
1         \n\nThe family of Dashwood had long been settl...
2         \n\nMrs. John Dashwood now installed herself m...
3         \n\nMrs. Dashwood remained at Norland several ...
4         \n\n"What a pity it is, Elinor," said Marianne...
5         \n\nNo sooner was her answer dispatched, than ...
6         \n\nThe first part of their journey was perfor...
7         \n\nBarton Park was about half a mile from the...
8         \n\nMrs. Jennings was a widow with an ample jo...
9         \n\nThe Dashwoods were now settled at Barton w...
10        \n\nMarianne's preserver, as Margaret, with mo...

In [83]:
CHAPS['chap_str'] = CHAPS.chap_str.str.strip()

In [84]:
CHAPS

chap_str
chap_num                                                   
1         The family of Dashwood had long been settled i...
2         Mrs. John Dashwood now installed herself mistr...
3         Mrs. Dashwood remained at Norland several mont...
4         "What a pity it is, Elinor," said Marianne, "t...
5         No sooner was her answer dispatched, than Mrs....
6         The first part of their journey was performed ...
7         Barton Park was about half a mile from the cot...
8         Mrs. Jennings was a widow with an ample jointu...
9         The Dashwoods were now settled at Barton with ...
10        Marianne's preserver, as Margaret, with more e...
11        Little had Mrs. Dashwood or her daughters imag...
12        As Elinor and Marianne were walking together t...
13        Their intended excursion to Whitwell turned ou...
14        The sudden termination of Colonel Brandon's vi...
15        Mrs. Dashwood's visit to Lady Middleton took p...
16        Marianne would have thought herself very inexc...
17        Mrs. Dashwood was surprised only for a moment ...
18        Elinor saw, with great uneasiness the low spir...
19        Edward remained a week at the cottage; he was ...
20        As the Miss Dashwoods entered the drawing-room...
21        The Palmers returned to Cleveland the next day...
22        Marianne, who had never much toleration for an...
23        However small Elinor's general dependence on L...
24        In a firm, though cautious tone, Elinor thus b...
25        Though Mrs. Jennings was in the habit of spend...
26        Elinor could not find herself in the carriage ...
27        "If this open weather holds much longer," said...
28        Nothing occurred during the next three or four...
29        Before the house-maid had lit their fire the n...
30        Mrs. Jennings came immediately to their room o...
31        From a night of more sleep than she had expect...
32        When the particulars of this conversation were...
33        After some opposition, Marianne yielded to her...
34        Mrs. John Dashwood had so much confidence in h...
35        Elinor's curiosity to see Mrs. Ferrars was sat...
36        Within a few days after this meeting, the news...
37        Mrs. Palmer was so well at the end of a fortni...
38        Mrs. Jennings was very warm in her praise of E...
39        The Miss Dashwoods had now been rather more th...
40        "Well, Miss Dashwood," said Mrs. Jennings, sag...
41        Edward, having carried his thanks to Colonel B...
42        One other short call in Harley Street, in whic...
43        Marianne got up the next morning at her usual ...
44        Elinor, starting back with a look of horror at...
45        Elinor, for some time after he left her, for s...
46        Marianne's illness, though weakening in its ki...
47        Mrs. Dashwood did not hear unmoved the vindica...
48        Elinor now found the difference between the ex...
49        Unaccountable, however, as the circumstances o...
50        After a proper resistance on the part of Mrs. ...

## Split chapters into paragraphs

In [85]:
para_pat = r'\n\n+'

In [86]:
PARAS = CHAPS['chap_str'].str.split(para_pat, expand=True).stack()\
    .to_frame('para_str').sort_index()
PARAS.index.names = OHCO[:2]

In [87]:
PARAS.head()

para_str
chap_num para_num                                                   
1        0         The family of Dashwood had long been settled i...
         1         By a former marriage, Mr. Henry Dashwood had o...
         2         The old gentleman died: his will was read, and...
         3         Mr. Dashwood's disappointment was, at first, s...
         4         His son was sent for as soon as his danger was...

In [88]:
PARAS['para_str'] = PARAS['para_str'].str.replace(r'\n', ' ', regex=True)
PARAS['para_str'] = PARAS['para_str'].str.strip()
PARAS = PARAS[~PARAS['para_str'].str.match(r'^\s*$')]

In [89]:
PARAS.head()

para_str
chap_num para_num                                                   
1        0         The family of Dashwood had long been settled i...
         1         By a former marriage, Mr. Henry Dashwood had o...
         2         The old gentleman died: his will was read, and...
         3         Mr. Dashwood's disappointment was, at first, s...
         4         His son was sent for as soon as his danger was...

## Split paragraphs into sentences


In [90]:
sent_pat = r'[.?!;:]+'
SENTS = PARAS['para_str'].str.split(sent_pat, expand=True).stack()\
    .to_frame('sent_str')
SENTS.index.names = OHCO[:3]

In [91]:
SENTS = SENTS[~SENTS['sent_str'].str.match(r'^\s*$')] 
SENTS.sent_str = SENTS.sent_str.str.strip()

In [92]:
SENTS.head(10)

sent_str
chap_num para_num sent_num                                                   
1        0        0         The family of Dashwood had long been settled i...
                  1         Their estate was large, and their residence wa...
                  2         The late owner of this estate was a single man...
                  3         But her death, which happened ten years before...
                  4         for to supply her loss, he invited and receive...
                  5         Henry Dashwood, the legal inheritor of the Nor...
                  6         In the society of his nephew and niece, and th...
                  7                      His attachment to them all increased
                  8                              The constant attention of Mr
                  9                                                   and Mrs

In [93]:
SENTS.sample(10)

sent_str
chap_num para_num sent_num                                                   
41       19       5           Ferrars can never forget that Edward is her son
13       10       1         It came from town, and is merely a letter of b...
28       16       4         Absence might have weakened his regard, and co...
33       40       1                                                  Jennings
37       20       1                                                         "
27       11       1         he came to look at Marianne and talk to Elinor...
22       34       2         She returned it almost instantly, acknowledgin...
37       26       3         Now, I can think and speak of it with little e...
40       50       1                                               " cried Mrs
29       68       4                                                         "

## Split sentences into tokens

In [113]:
token_pat = r"[\s',-]+"
TOKENS_ss = SENTS['sent_str'].str.split(token_pat, expand=True).stack()\
    .to_frame('token_str')
TOKENS_ss

token_str
chap_num para_num sent_num                
1        0        0        0           The
                           1        family
                           2            of
                           3      Dashwood
                           4           had
...                                    ...
50       22       0        8           and
                           9   Sensibility
                           10           by
                           11         Jane
                           12       Austen

[122882 rows x 1 columns]

In [95]:
TOKENS_ss.index.names = OHCO[:4]

In [96]:
TOKENS_ss

token_str
chap_num para_num sent_num token_num             
1        0        0        0                  The
                           1               family
                           2                   of
                           3             Dashwood
                           4                  had
...                                           ...
50       22       0        8                  and
                           9          Sensibility
                           10                  by
                           11                Jane
                           12              Austen

[122882 rows x 1 columns]

## Import Persuasion and Combine 

In [97]:
TOKENS_p = pd.read_csv(f"{data_home}/austen-persuasion.csv", index_col=OHCO)
TOKENS_p

token_str    term_str
chap_num para_num sent_num token_num                        
1        0        0        0                 Sir         sir
                           1              Walter      walter
                           2              Elliot      elliot
                           3                  of          of
                           4            Kellynch    kellynch
...                                          ...         ...
24       13       0        6                  of          of
                           7          Persuasion  persuasion
                           8                  by          by
                           9                Jane        jane
                           10             Austen      austen

[85014 rows x 2 columns]

In [98]:
TOKENS = pd.concat([TOKENS_ss,TOKENS_p], axis = 0, keys = [0,1], names = ['book'])

###  1. How many raw tokens are in the combined data frame?

In [99]:
TOKENS.shape

(207896, 2)

There are 207896 raw tokens in the combined data frame.

### 2. How many distinct terms are there in the combined data frame (i.e. how big is the vocabulary)?

### Extract Vocabulary

In [100]:
TOKENS['term_str'] = TOKENS.token_str.replace(r'[\W_]+', '', regex=True).str.lower()
VOCAB = TOKENS.term_str.value_counts().to_frame('n').reset_index().rename(columns={'index':'term_str'})
VOCAB.index.name = 'term_id'

In [101]:
VOCAB

term_str     n
term_id                     
0                  the  7435
1                   to  6923
2                  and  6290
3                   of  6146
4                  her  3747
...                ...   ...
8234     unconquerable     1
8235          outgrown     1
8236      prosperously     1
8237           nominal     1
8238             finis     1

[8239 rows x 2 columns]

In [102]:
VOCAB.shape

(8239, 2)

There are 8,239 distinct terms in the combined data frame.

### 3. How many more terms does the vocabulary of Sense and Sensibility have than that of Persuasion?

In [103]:
TOKENS_ss['term_str'] = TOKENS_ss.token_str.replace(r'[\W_]+', '', regex=True).str.lower()
VOCAB_ss = TOKENS_ss.term_str.value_counts().to_frame('n').reset_index().rename(columns={'index':'term_str'})
VOCAB_ss.index.name = 'term_id'
VOCAB_ss

term_str     n
term_id                   
0                 to  4115
1                the  4105
2                 of  3574
3                and  3490
4                her  2543
...              ...   ...
6275          prefer     1
6276      dissolving     1
6277           beset     1
6278     effectually     1
6279          austen     1

[6280 rows x 2 columns]

In [104]:
TOKENS_p['term_str'] = TOKENS_p.token_str.replace(r'[\W_]+', '', regex=True).str.lower()
VOCAB_p = TOKENS_p.term_str.value_counts().to_frame('n').reset_index().rename(columns={'index':'term_str'})
VOCAB_p.index.name = 'term_id'
VOCAB_p

term_str     n
term_id                   
0                the  3330
1                 to  2808
2                and  2800
3                 of  2572
4                  a  1595
...              ...   ...
5755           reins     1
5756     judiciously     1
5757             rut     1
5758            dung     1
5759          austen     1

[5760 rows x 2 columns]

In [105]:
(VOCAB_ss.shape[0]) - (VOCAB_p.shape[0])

520

Sense and Sensibility has 520 more unique terms than Persuasion.

### 4. What is the average number of tokens, rounded to an integer, per chapter in the corpus?

In [106]:
def gather(ohco_level):
    global TOKENS
    level_name = OHCO[ohco_level-1].split('_')[0]
    df = TOKENS.groupby(OHCO[:ohco_level])\
        .token_str.apply(lambda x: x.str.cat(sep=' '))\
        .to_frame(f"{level_name}_str")
    return df

In [107]:
TOKENS.groupby(['book','chap_num']).count().mean()

token_str    2807.918919
term_str     2807.918919
dtype: float64

The average number of tokens per chapter in the corpus is 2808

### 5. What is the average number of tokens, rounded to an integer, per paragraph in the corpus?

In [108]:
TOKENS.groupby(['book','chap_num','para_num']).count().mean()

token_str    73.709117
term_str     73.709117
dtype: float64

The average number of tokens per paragraph in the corpus is 74

In [109]:
TOKENS.to_csv(csv_file)

In [110]:
TOKENS_ss.to_csv(f"{output_dir}/austen-sense-and-sensibility.csv")